{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "9173ccb6-39c0-435b-ae7a-a5afe514203a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import gc_fraction, MeltingTemp as mt\n",
    "from collections import Counter\n",
    "import random\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "592c294f-ed6e-4e2b-b73c-86954cfd81cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate as many umis as possible\n",
    "# define the length\n",
    "umilen = 16\n",
    "\n",
    "# function to create umis\n",
    "def create_umi(length = umilen):\n",
    "    return ''.join(random.choices('ACGT', k=length))\n",
    "\n",
    "# for a sequence of 16 nucleotides, the average count for nucleotide must be 4\n",
    "# on average, counts should be 4!\n",
    "# we can set 1 as the highest deviation from the average that we accept for the umi to be \"equilibrated\"\n",
    "threshold =  1\n",
    "\n",
    "# function to check if nucleotides are in equilibrium\n",
    "def count_nns(umi, threshold):\n",
    "    counts = Counter(umi).values()\n",
    "    mean_count = sum(counts) / len(counts)\n",
    "    \n",
    "    # calculate deviations\n",
    "    devs = [abs(count - mean_count) for count in counts]\n",
    "    # check that all counts are within the deviation threshold\n",
    "    if all(dev <= threshold for dev in devs):\n",
    "        return umi\n",
    "        \n",
    "random.seed(666)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66b85b2c-eb81-4fd6-a82e-ed17b69eaf6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we have randomly chosen umis that are roughly equilibrated\n",
    "# now we need to check that the umi doesn't have more than N of the same nucleotide together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "27459194-bbe6-40f4-9c50-70cec5ebaa70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# homopolymers no longer than 0 == no homopolymers\n",
    "N = 0\n",
    "# number of homopolymers threshold\n",
    "P = 0\n",
    "\n",
    "def check_homopolymers(umiseq, N, P):\n",
    "    # create empty variables to store nucleotides\n",
    "    current_nn = None\n",
    "    conseq = 0 # initiate the count of nucleotides from zero\n",
    "    homseq = 0 # count for homopolymers\n",
    "    longest = 0 # keeps track of the longest homopolymer\n",
    "\n",
    "    # iterate through the umi sequence\n",
    "    for nn in umiseq:\n",
    "        if nn == current_nn:\n",
    "            conseq += 1 # add one to the count if subsequent nucleotides after current are the same\n",
    "            homseq += 1\n",
    "            if conseq >= longest:\n",
    "                longest = conseq\n",
    "        else:\n",
    "            # replace the variable with the next nucleotide if it is different\n",
    "            conseq = 0 # resets the nucleotide count\n",
    "        current_nn = nn  # Change to a new nucleotide\n",
    "\n",
    "    # check if the homopolymer count exceeds the N threshold\n",
    "    if homseq > P: # has at least one homopolymer\n",
    "        return False  # Homopolymer of length greater than N found\n",
    "    # filter by longest homopolymer\n",
    "    elif longest > N:\n",
    "        return False\n",
    "    else:\n",
    "        return umiseq  # No homopolymer longer than N found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "deb31f59-d4cc-4ea1-b4ca-ece42fbe7e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now, we need to check that the sequence is not complementary or within the target sequence \n",
    "\n",
    "def check_seq_match(umi, target):\n",
    "    rev_target = Seq(target)[::-1]  # reverse target sequence\n",
    "    rev_comp_target = Seq(target).reverse_complement()  # reverse complement target\n",
    "\n",
    "    # check the sequences\n",
    "    if str(umi) == str(target):\n",
    "        return False, 'UMI is the same as the target sequence'\n",
    "    elif str(umi) == str(rev_target):\n",
    "        return False, 'UMI is the reverse of the target sequence'\n",
    "    elif str(umi) == str(rev_comp_target):\n",
    "        return False, 'UMI is the reverse complement of the target sequence'\n",
    "\n",
    "    return umi, 'UMI is unique and does not match the target sequence'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0453b03-39b2-4ddf-8156-aea56a3baffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# but we also want to check for partial matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "f799c10f-07ad-46c6-9fde-2bd0ef542ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_partial_matches(umi, target, min_match = 6):\n",
    "    # Generate reverse and reverse complement of the target sequence\n",
    "    rev_target = Seq(target)[::-1]  # Reverse of the target sequence\n",
    "    rev_comp_target = Seq(target).reverse_complement()  # Reverse complement using BioPython\n",
    "    \n",
    "    # Function to check for partial matches in a given target sequence\n",
    "    def find_partial_matches(umi, target, min_match):\n",
    "        umi_len = len(str(umi))\n",
    "        target_len = len(str(target))\n",
    "\n",
    "        # function to iterate through nucleotides from zero to the target's length minus match length plus one\n",
    "        # that is the last kmer of length min_match + 1\n",
    "        for i in range(target_len - min_match + 1):\n",
    "            # iterates through nucleotides from the first kmer to the length of the umi plus +\n",
    "            for j in range(min_match, umi_len + 1):\n",
    "                # checks if umi in target and vice versa\n",
    "                if target[i:i+j] in umi or umi in target[i:i+j]:\n",
    "                    return 'Partial match', target[i:i+j] # there's a match\n",
    "        # this umi is returning the sequence\n",
    "        return 'No match of length %s'%(min_match), 'comparing: %s'%(target[i:i+j])\n",
    "    \n",
    "    # check partial matches of umi on the target, reverse, and reverse complement\n",
    "    # using the iterating function above\n",
    "    for seq in [target, str(rev_target), str(rev_comp_target)]:\n",
    "        umitemp, match_sequence = find_partial_matches(umi, seq, min_match)\n",
    "        # if there is a match, the match_sequence string will contain text and the matching sequences\n",
    "        # thus, it would be longer than the minimum match length allowed\n",
    "        if len(match_sequence) > min_match:\n",
    "            return False, f'Partial match found: {match_sequence} in sequence {seq}'\n",
    "    return umi, 'No significant partial matches found'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "ec51c81c-63e4-4508-b0c8-1d347d6133a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "TACGTACGTACGTACG\n",
      "(False, 'UMI is the same as the target sequence')\n",
      "('TACGTACGTACGTACG', 'UMI is unique and does not match the target sequence')\n",
      "(False, 'Partial match found: comparing: CCAAA in sequence GCTAGGGTTTCCCAAA')\n",
      "(False, 'Partial match found: comparing: GATCG in sequence AAACCCTTTGGGATCG')\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "# test the tree functions\n",
    "test = 'AAACCCTTTGGGATCG'\n",
    "test1 = 'TACGTACGTACGTACG'\n",
    "\n",
    "target = 'AAACCCTTTGGGATCG'\n",
    "print(check_homopolymers(test, N, P))\n",
    "print(check_homopolymers(test1, N, P))\n",
    "\n",
    "print(check_seq_match(test,target))\n",
    "print(check_seq_match(test1,target))\n",
    "\n",
    "print(check_partial_matches(test,target, 5))\n",
    "print(check_partial_matches(test1,target, 5))\n",
    "\n",
    "umipass, statement = check_partial_matches(test1,target, 5)\n",
    "print(umipass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "17b2ac76-6cf8-401b-82f4-2f179216c7c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding UMIs with no full match:\n",
      "Remaining UMIs 10\n",
      "['CTGCTGCAGAATACAT', 'TACACTGCGCAGCTAT', 'CAGGAGTAACTACTGC', 'GTACGCCCAGCATATG', 'CGTATCTAGGTAACAT', 'TGATGGACTCACCAGC', 'GTCTGCGTGATTAACA', 'GAACTAGCTGCAGGCT', 'TATGCTAGACAGGTCA', 'GAGTACATCGCTCAGT'] \n",
      "\n",
      "UMIs with no full match\n",
      "Remaining UMIs 2\n",
      "['CTGCTGCAGAATACAT', 'TGATGGACTCACCAGC'] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "target = 'CACACCTCGGTGTGAAGCAAATGATTGTTGCTATTAACAAGATGGACGACAAATCTGTCAACTGGGCACAATCTAGGTACGATGAAATAGTTAAGGAAGTATCCTCCTTTGTCAAGAAGATCGGCTACAACCCTGAGAAGATCCCGTTCGTCCCCATTTCTGGTTGGCACGGAGATAACATGCTCGAGAAGTCATCTAACCTCTCATGGTATAAAGGCCCCACATTACTCGAGGCCCTCGACTCTGTGTCAGAACCCAAGAGACCAACGGAAAAGCCCCTCCGAATTCCCCTTCAAGATGTTTACAAGATTGGAGGTATTGGAACTGTGCCTGTAGGCCGTGTTGAGACTGGGGTTCTCAAACCAGGAATGAACGTTACTTTCTCCCCTGCTGGTTTGACCACTGAAGTTAAGTCTGTTGAGATGCACCACGTCTCTCTCCCTGAGGCTGTCCCAGGTGATAACGTTGGTTTCAATGTCAAGAATCTGTCAGTTAAGGATATTCGTCGTGGTATGGTTGCTGGTGATGCCAAGAATGATCCCCCTCAAGAGACTGAAGATTTTAATGCCCAAGTTATTATTCTCAATCACCCTGGTCAGATCCATGCAGGATATGCCCCAGTGCTTGATTGTCACACTGCTCATATCGCCTGTAAGTTCAGCGAGATTCTCTCCAAAGTAGATCGTAGATCTGGTCAAGAGACTGAGGCTGCCCCTAAGAATATCAAGAACGGAGATGCCCGCCATAGTTAAACTCACTCCCTCCCAAGGCCCATGTGTGTGGAGTCTTTCTCTGATTACCCACCCCCTTGG'\n",
    "testlist = ['CTGCTGCAGAATACAT', 'TACACTGCGCAGCTAT', 'CAGGAGTAACTACTGC', 'GTACGCCCAGCATATG', 'CGTATCTAGGTAACAT', 'TGATGGACTCACCAGC', 'GTCTGCGTGATTAACA', 'GAACTAGCTGCAGGCT', 'TATGCTAGACAGGTCA', 'GAGTACATCGCTCAGT']\n",
    "\n",
    "newlist = []\n",
    "for umi in testlist:\n",
    "    check_seq_match(umi, target)\n",
    "    umibool, statement = check_seq_match(umi, target)\n",
    "    # print('step1', umibool, statement)\n",
    "    if umibool != False: \n",
    "        # print('step2', umibool, statement)\n",
    "        if umibool not in newlist and umibool != True:\n",
    "            # print('umibool', umibool)\n",
    "            newlist.append(umibool)\n",
    "            \n",
    "print('Finding UMIs with no full match:')\n",
    "print('Remaining UMIs %s'%(len(newlist)))\n",
    "print(newlist[0:10], '\\n')\n",
    "\n",
    "\n",
    "newlist1 = []\n",
    "for umi in newlist:\n",
    "    umibool, statement = check_partial_matches(umi, target, min_match = 6)\n",
    "    # print('step1', umibool, statement)\n",
    "    if umibool != False:\n",
    "        # print('step2', umibool, statement)\n",
    "        if umibool not in newlist1:\n",
    "            # print('umibool', umibool)\n",
    "            newlist1.append(umibool)\n",
    "            \n",
    "print('UMIs with no partial match')\n",
    "print('Remaining UMIs %s'%(len(newlist1)))\n",
    "print(newlist1[0:10], '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "8a0cba90-9754-44f8-b9bc-1b27a14a9047",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw UMIs created: 1000001\n",
      "['TTGATCGAGCCACTGG', 'GGTAAAACACGTCGTT', 'ATGAGTAGATCCCCGA', 'ACATAAGTTCAGGTCC', 'CTAAAGGGTCTTCGTG', 'TAATCCGCAAGGGAGT', 'TCGTCTAGCTCGAAAA', 'TCAACGGCTAGCGGTC', 'AATACTCCCTGAGGAG', 'TATCAGGCCGAAACTC'] \n",
      "\n",
      "Non-homopolymer UMIs with homopolymer number threshold 2 and nucleotide count threshold 2\n",
      "Remaining UMIs 351031\n",
      "['CGATATGAGCAATGCC', 'TATCATGATCTCAGGA', 'GCACTCATGTGAGCAG', 'AGACTCTGACTTGTAA', 'GGACGTGTACATCATG', 'ATAGCTGCTGTTCACA', 'ACGCGTCTGAAGTAGT', 'GAGGTGCAGATACTTC', 'TTAGTCTAGTCAAGCA', 'AGACCGGTACTAGATG'] \n",
      "\n",
      "UMIs with no full match\n",
      "Remaining UMIs 351031\n",
      "['CGATATGAGCAATGCC', 'TATCATGATCTCAGGA', 'GCACTCATGTGAGCAG', 'AGACTCTGACTTGTAA', 'GGACGTGTACATCATG', 'ATAGCTGCTGTTCACA', 'ACGCGTCTGAAGTAGT', 'GAGGTGCAGATACTTC', 'TTAGTCTAGTCAAGCA', 'AGACCGGTACTAGATG'] \n",
      "\n",
      "UMIs with no partial match\n",
      "Remaining UMIs 142435\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['TATCATGATCTCAGGA',\n",
       " 'GCACTCATGTGAGCAG',\n",
       " 'AGACTCTGACTTGTAA',\n",
       " 'ATAGCTGCTGTTCACA',\n",
       " 'TTAGTCTAGTCAAGCA',\n",
       " 'GACACACCGTTAGTGT',\n",
       " 'AACTCGTCATGTGGTA',\n",
       " 'TCATGTACCTGACGAC',\n",
       " 'AATGCATACTCGAGCG',\n",
       " 'CAGTGTGTCAACACGG']"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# implement the tree functions\n",
    "# assume the following target barcode in Fungi\n",
    "# TEF1α region Al33F (5′-GAYTTCATCAAGAACATGAT-′3) and Al33R (5′-GACGTTGAADCCRACRTTGTC-′3)\n",
    "# from https://cdnsciencepub.com/doi/10.1139/gen-2018-0083\n",
    "# create umis\n",
    "\n",
    "# Trichia persimili, a myxomicetes (group found in deadwood, doesnt mean this spp is saproxylic)\n",
    "# GenBank: FJ546692.1\n",
    "target = 'CACACCTCGGTGTGAAGCAAATGATTGTTGCTATTAACAAGATGGACGACAAATCTGTCAACTGGGCACAATCTAGGTACGATGAAATAGTTAAGGAAGTATCCTCCTTTGTCAAGAAGATCGGCTACAACCCTGAGAAGATCCCGTTCGTCCCCATTTCTGGTTGGCACGGAGATAACATGCTCGAGAAGTCATCTAACCTCTCATGGTATAAAGGCCCCACATTACTCGAGGCCCTCGACTCTGTGTCAGAACCCAAGAGACCAACGGAAAAGCCCCTCCGAATTCCCCTTCAAGATGTTTACAAGATTGGAGGTATTGGAACTGTGCCTGTAGGCCGTGTTGAGACTGGGGTTCTCAAACCAGGAATGAACGTTACTTTCTCCCCTGCTGGTTTGACCACTGAAGTTAAGTCTGTTGAGATGCACCACGTCTCTCTCCCTGAGGCTGTCCCAGGTGATAACGTTGGTTTCAATGTCAAGAATCTGTCAGTTAAGGATATTCGTCGTGGTATGGTTGCTGGTGATGCCAAGAATGATCCCCCTCAAGAGACTGAAGATTTTAATGCCCAAGTTATTATTCTCAATCACCCTGGTCAGATCCATGCAGGATATGCCCCAGTGCTTGATTGTCACACTGCTCATATCGCCTGTAAGTTCAGCGAGATTCTCTCCAAAGTAGATCGTAGATCTGGTCAAGAGACTGAGGCTGCCCCTAAGAATATCAAGAACGGAGATGCCCGCCATAGTTAAACTCACTCCCTCCCAAGGCCCATGTGTGTGGAGTCTTTCTCTGATTACCCACCCCCTTGG'\n",
    "\n",
    "umi_raw = []\n",
    "# to generate 96 forward and 96 reverse umis\n",
    "# and check for nucleotide composition balance\n",
    "# 4**16 is as many combinations as we can get =  4294967296\n",
    "# 100000 is the number of initial random tags, you can reduce or increase the number if it takes too long\n",
    "# or if no enough umis pass the test\n",
    "while len(umi_raw) <= 100000:\n",
    "    umi = create_umi()\n",
    "    umi_pass1 = count_nns(umi, threshold)\n",
    "    \n",
    "    # add the umi to the list if it is not already there\n",
    "    if umi_pass1 not in umi_raw and umi_pass1 != None:\n",
    "        umi_raw.append(umi_pass1)\n",
    "print('Raw UMIs created: %s'%(len(umi_raw)))\n",
    "print(umi_raw[0:10], '\\n')\n",
    "\n",
    "# check umis for homopolymers\n",
    "\n",
    "N = 2 # allow the longest homopolymer to have two nucleotides\n",
    "P = 2 # allow a max of one homopolymer\n",
    "\n",
    "umi_uniq = []\n",
    "\n",
    "for umi in umi_raw:\n",
    "    if check_homopolymers(umi, N, P) is False:\n",
    "        pass\n",
    "    else:\n",
    "        if umi not in umi_uniq:\n",
    "            umi_uniq.append(umi)\n",
    "print('Non-homopolymer UMIs with homopolymer number threshold %s and nucleotide count threshold %s'%(P,N))\n",
    "print('Remaining UMIs %s'%(len(umi_uniq)))\n",
    "print(umi_uniq[0:10], '\\n')\n",
    "\n",
    "# check for matches\n",
    "umi_nofmatch = []\n",
    "\n",
    "for umi in umi_uniq:\n",
    "    check_seq_match(umi, target)\n",
    "    umibool, statement = check_seq_match(umi, target)\n",
    "    if umibool != False: \n",
    "        if umibool not in umi_nofmatch and umibool != True:\n",
    "            umi_nofmatch.append(umibool)\n",
    "print('UMIs with no full match')\n",
    "print('Remaining UMIs %s'%(len(umi_nofmatch)))\n",
    "print(umi_nofmatch[0:10], '\\n')\n",
    "\n",
    "# check for partial matches\n",
    "umi_preselected = []\n",
    "\n",
    "for umi in umi_nofmatch:\n",
    "    umibool, statement = check_partial_matches(umi, target, min_match = 6)\n",
    "    if umibool != False:\n",
    "        if umibool not in umi_preselected:\n",
    "            umi_preselected.append(umibool)\n",
    "            \n",
    "print('UMIs with no partial match')\n",
    "print('Remaining UMIs %s'%(len(umi_preselected))) # number of umis that pass all tests\n",
    "umi_preselected[0:10] # prints the first 10 in the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "ef1fbbc2-9f35-41cf-a521-15a419d8a8d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with forward primers\n",
      "done with reverse primers\n",
      "Umis generated under the thresholds: 192\n",
      "Of which 96 are forward and 96 are reverse\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense</th>\n",
       "      <th>padd_seq</th>\n",
       "      <th>umi_seq</th>\n",
       "      <th>primer_seq</th>\n",
       "      <th>full_primer</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GCGATTGCCTGCACAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGGCGATTGCCTGCACAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.610557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GTCGCCCAGCATGTGA</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGGTCGCCCAGCATGTGAGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.160343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GACGGACGCAGTCTCT</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGGACGGACGCAGTCTCTGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.667593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GCGCGACTTGAGATCC</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGGCGCGACTTGAGATCCGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.163848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>CGAAGACTCGCGTCGT</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGCGAAGACTCGCGTCGTGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.368894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GTACGGCAGACTGTCC</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGGTACGGCAGACTGTCCGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.678127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>CGAGTCTTCAGCGACG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGCGAGTCTTCAGCGACGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.109110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>CGGCACAGAGTTCTGC</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGCGGCACAGAGTTCTGCGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.543285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GCATGAGGCACTCGTC</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGGCATGAGGCACTCGTCGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.152573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>forward</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>CTCGCTGTAGAGCCAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>GGTAGCTCGCTGTAGAGCCAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.728494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sense padd_seq           umi_seq            primer_seq  \\\n",
       "0  forward    GGTAG  GCGATTGCCTGCACAG  GAYTTCATCAAGAACATGAT   \n",
       "1  forward    GGTAG  GTCGCCCAGCATGTGA  GAYTTCATCAAGAACATGAT   \n",
       "2  forward    GGTAG  GACGGACGCAGTCTCT  GAYTTCATCAAGAACATGAT   \n",
       "3  forward    GGTAG  GCGCGACTTGAGATCC  GAYTTCATCAAGAACATGAT   \n",
       "4  forward    GGTAG  CGAAGACTCGCGTCGT  GAYTTCATCAAGAACATGAT   \n",
       "5  forward    GGTAG  GTACGGCAGACTGTCC  GAYTTCATCAAGAACATGAT   \n",
       "6  forward    GGTAG  CGAGTCTTCAGCGACG  GAYTTCATCAAGAACATGAT   \n",
       "7  forward    GGTAG  CGGCACAGAGTTCTGC  GAYTTCATCAAGAACATGAT   \n",
       "8  forward    GGTAG  GCATGAGGCACTCGTC  GAYTTCATCAAGAACATGAT   \n",
       "9  forward    GGTAG  CTCGCTGTAGAGCCAG  GAYTTCATCAAGAACATGAT   \n",
       "\n",
       "                                 full_primer  gc_content       temp  \n",
       "0  GGTAGGCGATTGCCTGCACAGGAYTTCATCAAGAACATGAT        47.5  66.610557  \n",
       "1  GGTAGGTCGCCCAGCATGTGAGAYTTCATCAAGAACATGAT        47.5  66.160343  \n",
       "2  GGTAGGACGGACGCAGTCTCTGAYTTCATCAAGAACATGAT        47.5  65.667593  \n",
       "3  GGTAGGCGCGACTTGAGATCCGAYTTCATCAAGAACATGAT        47.5  66.163848  \n",
       "4  GGTAGCGAAGACTCGCGTCGTGAYTTCATCAAGAACATGAT        47.5  66.368894  \n",
       "5  GGTAGGTACGGCAGACTGTCCGAYTTCATCAAGAACATGAT        47.5  65.678127  \n",
       "6  GGTAGCGAGTCTTCAGCGACGGAYTTCATCAAGAACATGAT        47.5  66.109110  \n",
       "7  GGTAGCGGCACAGAGTTCTGCGAYTTCATCAAGAACATGAT        47.5  66.543285  \n",
       "8  GGTAGGCATGAGGCACTCGTCGAYTTCATCAAGAACATGAT        47.5  66.152573  \n",
       "9  GGTAGCTCGCTGTAGAGCCAGGAYTTCATCAAGAACATGAT        47.5  65.728494  "
      ]
     },
     "execution_count": 261,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finally, calculate GC content and annealing temperature\n",
    "# for the remaining 142435 umis (we need 96*2)\n",
    "# we can check the first 1000 or so umis\n",
    "# we will select those with compatible temperatures and good CG content\n",
    "\n",
    "padd = 'GGTAG'\n",
    "primer_forward = 'GAYTTCATCAAGAACATGAT'\n",
    "primer_reverse = 'GACGTTGAADCCRACRTTGTC'\n",
    "gc_min = 46\n",
    "gc_max = 54\n",
    "# based on https://www.tandfonline.com/doi/full/10.1080/15572536.2006.11832842#d1e421 - FUNGI!\n",
    "temp_min = 55\n",
    "temp_max = 67\n",
    "\n",
    "# how many tagged primers to generate\n",
    "# e.g. if num_primers = 96, it will generate 96 forward and 96 reverse\n",
    "num_primers = 96\n",
    "\n",
    "# creates the table to store information\n",
    "tagged_primers = pd.DataFrame(columns = ['sense','padd_seq','umi_seq','primer_seq','full_primer','gc_content','temp'])\n",
    "\n",
    "# to make sure no same umi is used for two primer pairs:\n",
    "umi_track = []\n",
    "\n",
    "# initiate counters to generate just the primers necessary\n",
    "f_count = 0\n",
    "for umi in umi_preselected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = padd + umi + primer_forward\n",
    "        # print(fullseq)\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        # print(gc_content)\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        # print(temp)\n",
    "        sense = 'forward' # python is zero indexed\n",
    "\n",
    "        # add thresholds if you wanna keep padd+umi+primer temperatures that overlap the\n",
    "        # PCR protocol annealing temperatures already\n",
    "        # and to keep the GC content close to 50% - second filter for that\n",
    "        # greater than or equal AND lesser then or equal to\n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, padd, umi, primer, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            # add to the counters\n",
    "            f_count += 1\n",
    "            # print('passed %s'%(f_count))\n",
    "            if f_count == num_primers:\n",
    "                break\n",
    "print('done with forward primers')\n",
    "        \n",
    "r_count = 0\n",
    "for umi in umi_preselected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = padd + umi + primer\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'reverse' # python is zero indexed\n",
    "        \n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, padd, umi, primer, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            r_count += 1\n",
    "            if r_count == num_primers:\n",
    "                break\n",
    "    \n",
    "print('done with reverse primers')\n",
    "            \n",
    "print('Umis generated under the thresholds: %s'%(len(tagged_primers)))\n",
    "print('Of which %s are forward and %s are reverse'%(len(tagged_primers[tagged_primers['sense'] == 'forward']),\n",
    "                                                    len(tagged_primers[tagged_primers['sense'] == 'reverse'])))\n",
    "tagged_primers.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "7bf004ce-91ce-4549-af48-ee3777f63755",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file!\n",
    "# TEF1α region Al33F (5′-GAYTTCATCAAGAACATGAT-′3) and Al33R (5′-GACGTTGAADCCRACRTTGTC-′3)\n",
    "tagged_primers.to_csv('./TEFa_Al33F-Al33R_tagged.csv', sep = '\\t', index_label = 'index')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
