{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "9173ccb6-39c0-435b-ae7a-a5afe514203a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import gc_fraction, MeltingTemp as mt\n",
    "from collections import Counter\n",
    "from seqfold import dg, fold\n",
    "import random\n",
    "import re\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "592c294f-ed6e-4e2b-b73c-86954cfd81cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate as many umis as possible\n",
    "# define the length\n",
    "umilen = 16\n",
    "\n",
    "# function to create umis\n",
    "def create_umi(length = umilen):\n",
    "    return ''.join(random.choices('ACGT', k=length))\n",
    "\n",
    "# for a sequence of 16 nucleotides, the average count for nucleotide must be 4\n",
    "# on average, counts should be 4!\n",
    "# we can set 1 as the highest deviation from the average that we accept for the umi to be \"equilibrated\"\n",
    "threshold =  1\n",
    "\n",
    "# function to check if nucleotides are in equilibrium\n",
    "def count_nns(umi, threshold):\n",
    "    counts = Counter(umi).values()\n",
    "    mean_count = sum(counts) / len(counts)\n",
    "    \n",
    "    # calculate deviations\n",
    "    devs = [abs(count - mean_count) for count in counts]\n",
    "    # check that all counts are within the deviation threshold\n",
    "    if all(dev <= threshold for dev in devs):\n",
    "        return umi\n",
    "        \n",
    "random.seed(666)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66b85b2c-eb81-4fd6-a82e-ed17b69eaf6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we have randomly chosen umis that are roughly equilibrated\n",
    "# now we need to check that the umi doesn't have more than N of the same nucleotide together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "27459194-bbe6-40f4-9c50-70cec5ebaa70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# homopolymers no longer than 0 == no homopolymers\n",
    "N = 0\n",
    "# number of homopolymers threshold\n",
    "P = 0\n",
    "\n",
    "def check_homopolymers(umiseq, N, P):\n",
    "    # create empty variables to store nucleotides\n",
    "    current_nn = None\n",
    "    conseq = 0 # initiate the count of nucleotides from zero\n",
    "    homseq = 0 # count for homopolymers\n",
    "    longest = 0 # keeps track of the longest homopolymer\n",
    "\n",
    "    # iterate through the umi sequence\n",
    "    for nn in umiseq:\n",
    "        if nn == current_nn:\n",
    "            conseq += 1 # add one to the count if subsequent nucleotides after current are the same\n",
    "            homseq += 1\n",
    "            if conseq >= longest:\n",
    "                longest = conseq\n",
    "        else:\n",
    "            # replace the variable with the next nucleotide if it is different\n",
    "            conseq = 0 # resets the nucleotide count\n",
    "        current_nn = nn  # Change to a new nucleotide\n",
    "\n",
    "    # check if the homopolymer count exceeds the N threshold\n",
    "    if homseq > P: # has at least one homopolymer\n",
    "        return False  # Homopolymer of length greater than N found\n",
    "    # filter by longest homopolymer\n",
    "    elif longest > N:\n",
    "        return False\n",
    "    else:\n",
    "        return umiseq  # No homopolymer longer than N found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "deb31f59-d4cc-4ea1-b4ca-ece42fbe7e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now, we need to check that the sequence is not complementary or within the target sequence \n",
    "\n",
    "def check_seq_match(umi, target):\n",
    "    rev_target = Seq(target)[::-1]  # reverse target sequence\n",
    "    rev_comp_target = Seq(target).reverse_complement()  # reverse complement target\n",
    "\n",
    "    # check the sequences\n",
    "    if str(umi) == str(target):\n",
    "        return False, 'UMI is the same as the target sequence'\n",
    "    elif str(umi) == str(rev_target):\n",
    "        return False, 'UMI is the reverse of the target sequence'\n",
    "    elif str(umi) == str(rev_comp_target):\n",
    "        return False, 'UMI is the reverse complement of the target sequence'\n",
    "\n",
    "    return umi, 'UMI is unique and does not match the target sequence'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0453b03-39b2-4ddf-8156-aea56a3baffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# but we also want to check for partial matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "f799c10f-07ad-46c6-9fde-2bd0ef542ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_partial_matches(umi, target, min_match = 6):\n",
    "    # Generate reverse and reverse complement of the target sequence\n",
    "    rev_target = Seq(target)[::-1]  # Reverse of the target sequence\n",
    "    rev_comp_target = Seq(target).reverse_complement()  # Reverse complement using BioPython\n",
    "    \n",
    "    # Function to check for partial matches in a given target sequence\n",
    "    def find_partial_matches(umi, target, min_match):\n",
    "        umi_len = len(str(umi))\n",
    "        target_len = len(str(target))\n",
    "\n",
    "        # function to iterate through nucleotides from zero to the target's length minus match length plus one\n",
    "        # that is the last kmer of length min_match + 1\n",
    "        for i in range(target_len - min_match + 1):\n",
    "            # iterates through nucleotides from the first kmer to the length of the umi plus +\n",
    "            for j in range(min_match, umi_len + 1):\n",
    "                # checks if umi in target and vice versa\n",
    "                if target[i:i+j] in umi or umi in target[i:i+j]:\n",
    "                    return 'Partial match', target[i:i+j] # there's a match\n",
    "        # this umi is returning the sequence\n",
    "        return 'No match of length %s'%(min_match), 'comparing: %s'%(target[i:i+j])\n",
    "    \n",
    "    # check partial matches of umi on the target, reverse, and reverse complement\n",
    "    # using the iterating function above\n",
    "    for seq in [target, str(rev_target), str(rev_comp_target)]:\n",
    "        umitemp, match_sequence = find_partial_matches(umi, seq, min_match)\n",
    "        # if there is a match, the match_sequence string will contain text and the matching sequences\n",
    "        # thus, it would be longer than the minimum match length allowed\n",
    "        if len(match_sequence) > min_match:\n",
    "            return False, f'Partial match found: {match_sequence} in sequence {seq}'\n",
    "    return umi, 'No significant partial matches found'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "ec51c81c-63e4-4508-b0c8-1d347d6133a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "TACGTACGTACGTACG\n",
      "(False, 'UMI is the same as the target sequence')\n",
      "('TACGTACGTACGTACG', 'UMI is unique and does not match the target sequence')\n",
      "(False, 'Partial match found: comparing: CCAAA in sequence GCTAGGGTTTCCCAAA')\n",
      "(False, 'Partial match found: comparing: GATCG in sequence AAACCCTTTGGGATCG')\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "# test the tree functions\n",
    "test = 'AAACCCTTTGGGATCG'\n",
    "test1 = 'TACGTACGTACGTACG'\n",
    "\n",
    "target = 'AAACCCTTTGGGATCG'\n",
    "print(check_homopolymers(test, N, P))\n",
    "print(check_homopolymers(test1, N, P))\n",
    "\n",
    "print(check_seq_match(test,target))\n",
    "print(check_seq_match(test1,target))\n",
    "\n",
    "print(check_partial_matches(test,target, 5))\n",
    "print(check_partial_matches(test1,target, 5))\n",
    "\n",
    "umipass, statement = check_partial_matches(test1,target, 5)\n",
    "print(umipass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "17b2ac76-6cf8-401b-82f4-2f179216c7c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding UMIs with no full match:\n",
      "Remaining UMIs 10\n",
      "['CTGCTGCAGAATACAT', 'TACACTGCGCAGCTAT', 'CAGGAGTAACTACTGC', 'GTACGCCCAGCATATG', 'CGTATCTAGGTAACAT', 'TGATGGACTCACCAGC', 'GTCTGCGTGATTAACA', 'GAACTAGCTGCAGGCT', 'TATGCTAGACAGGTCA', 'GAGTACATCGCTCAGT'] \n",
      "\n",
      "UMIs with no full match\n",
      "Remaining UMIs 2\n",
      "['CTGCTGCAGAATACAT', 'TGATGGACTCACCAGC'] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "target = 'CACACCTCGGTGTGAAGCAAATGATTGTTGCTATTAACAAGATGGACGACAAATCTGTCAACTGGGCACAATCTAGGTACGATGAAATAGTTAAGGAAGTATCCTCCTTTGTCAAGAAGATCGGCTACAACCCTGAGAAGATCCCGTTCGTCCCCATTTCTGGTTGGCACGGAGATAACATGCTCGAGAAGTCATCTAACCTCTCATGGTATAAAGGCCCCACATTACTCGAGGCCCTCGACTCTGTGTCAGAACCCAAGAGACCAACGGAAAAGCCCCTCCGAATTCCCCTTCAAGATGTTTACAAGATTGGAGGTATTGGAACTGTGCCTGTAGGCCGTGTTGAGACTGGGGTTCTCAAACCAGGAATGAACGTTACTTTCTCCCCTGCTGGTTTGACCACTGAAGTTAAGTCTGTTGAGATGCACCACGTCTCTCTCCCTGAGGCTGTCCCAGGTGATAACGTTGGTTTCAATGTCAAGAATCTGTCAGTTAAGGATATTCGTCGTGGTATGGTTGCTGGTGATGCCAAGAATGATCCCCCTCAAGAGACTGAAGATTTTAATGCCCAAGTTATTATTCTCAATCACCCTGGTCAGATCCATGCAGGATATGCCCCAGTGCTTGATTGTCACACTGCTCATATCGCCTGTAAGTTCAGCGAGATTCTCTCCAAAGTAGATCGTAGATCTGGTCAAGAGACTGAGGCTGCCCCTAAGAATATCAAGAACGGAGATGCCCGCCATAGTTAAACTCACTCCCTCCCAAGGCCCATGTGTGTGGAGTCTTTCTCTGATTACCCACCCCCTTGG'\n",
    "testlist = ['CTGCTGCAGAATACAT', 'TACACTGCGCAGCTAT', 'CAGGAGTAACTACTGC', 'GTACGCCCAGCATATG', 'CGTATCTAGGTAACAT', 'TGATGGACTCACCAGC', 'GTCTGCGTGATTAACA', 'GAACTAGCTGCAGGCT', 'TATGCTAGACAGGTCA', 'GAGTACATCGCTCAGT']\n",
    "\n",
    "newlist = []\n",
    "for umi in testlist:\n",
    "    check_seq_match(umi, target)\n",
    "    umibool, statement = check_seq_match(umi, target)\n",
    "    # print('step1', umibool, statement)\n",
    "    if umibool != False: \n",
    "        # print('step2', umibool, statement)\n",
    "        if umibool not in newlist and umibool != True:\n",
    "            # print('umibool', umibool)\n",
    "            newlist.append(umibool)\n",
    "            \n",
    "print('Finding UMIs with no full match:')\n",
    "print('Remaining UMIs %s'%(len(newlist)))\n",
    "print(newlist[0:10], '\\n')\n",
    "\n",
    "\n",
    "newlist1 = []\n",
    "for umi in newlist:\n",
    "    umibool, statement = check_partial_matches(umi, target, min_match = 6)\n",
    "    # print('step1', umibool, statement)\n",
    "    if umibool != False:\n",
    "        # print('step2', umibool, statement)\n",
    "        if umibool not in newlist1:\n",
    "            # print('umibool', umibool)\n",
    "            newlist1.append(umibool)\n",
    "            \n",
    "print('UMIs with no partial match')\n",
    "print('Remaining UMIs %s'%(len(newlist1)))\n",
    "print(newlist1[0:10], '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "8a0cba90-9754-44f8-b9bc-1b27a14a9047",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw UMIs created: 1000001\n",
      "['TTGATCGAGCCACTGG', 'GGTAAAACACGTCGTT', 'ATGAGTAGATCCCCGA', 'ACATAAGTTCAGGTCC', 'CTAAAGGGTCTTCGTG', 'TAATCCGCAAGGGAGT', 'TCGTCTAGCTCGAAAA', 'TCAACGGCTAGCGGTC', 'AATACTCCCTGAGGAG', 'TATCAGGCCGAAACTC'] \n",
      "\n",
      "Non-homopolymer UMIs with homopolymer number threshold 2 and nucleotide count threshold 2\n",
      "Remaining UMIs 351031\n",
      "['CGATATGAGCAATGCC', 'TATCATGATCTCAGGA', 'GCACTCATGTGAGCAG', 'AGACTCTGACTTGTAA', 'GGACGTGTACATCATG', 'ATAGCTGCTGTTCACA', 'ACGCGTCTGAAGTAGT', 'GAGGTGCAGATACTTC', 'TTAGTCTAGTCAAGCA', 'AGACCGGTACTAGATG'] \n",
      "\n",
      "UMIs with no full match\n",
      "Remaining UMIs 351031\n",
      "['CGATATGAGCAATGCC', 'TATCATGATCTCAGGA', 'GCACTCATGTGAGCAG', 'AGACTCTGACTTGTAA', 'GGACGTGTACATCATG', 'ATAGCTGCTGTTCACA', 'ACGCGTCTGAAGTAGT', 'GAGGTGCAGATACTTC', 'TTAGTCTAGTCAAGCA', 'AGACCGGTACTAGATG'] \n",
      "\n",
      "UMIs with no partial match\n",
      "Remaining UMIs 142435\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['TATCATGATCTCAGGA',\n",
       " 'GCACTCATGTGAGCAG',\n",
       " 'AGACTCTGACTTGTAA',\n",
       " 'ATAGCTGCTGTTCACA',\n",
       " 'TTAGTCTAGTCAAGCA',\n",
       " 'GACACACCGTTAGTGT',\n",
       " 'AACTCGTCATGTGGTA',\n",
       " 'TCATGTACCTGACGAC',\n",
       " 'AATGCATACTCGAGCG',\n",
       " 'CAGTGTGTCAACACGG']"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# implement the tree functions\n",
    "# assume the following target barcode in Fungi\n",
    "# TEF1α region Al33F (5′-GAYTTCATCAAGAACATGAT-′3) and Al33R (5′-GACGTTGAADCCRACRTTGTC-′3)\n",
    "# from https://cdnsciencepub.com/doi/10.1139/gen-2018-0083\n",
    "# create umis\n",
    "\n",
    "# Trichia persimili, a myxomicetes (group found in deadwood, doesnt mean this spp is saproxylic)\n",
    "# GenBank: FJ546692.1\n",
    "target = 'CACACCTCGGTGTGAAGCAAATGATTGTTGCTATTAACAAGATGGACGACAAATCTGTCAACTGGGCACAATCTAGGTACGATGAAATAGTTAAGGAAGTATCCTCCTTTGTCAAGAAGATCGGCTACAACCCTGAGAAGATCCCGTTCGTCCCCATTTCTGGTTGGCACGGAGATAACATGCTCGAGAAGTCATCTAACCTCTCATGGTATAAAGGCCCCACATTACTCGAGGCCCTCGACTCTGTGTCAGAACCCAAGAGACCAACGGAAAAGCCCCTCCGAATTCCCCTTCAAGATGTTTACAAGATTGGAGGTATTGGAACTGTGCCTGTAGGCCGTGTTGAGACTGGGGTTCTCAAACCAGGAATGAACGTTACTTTCTCCCCTGCTGGTTTGACCACTGAAGTTAAGTCTGTTGAGATGCACCACGTCTCTCTCCCTGAGGCTGTCCCAGGTGATAACGTTGGTTTCAATGTCAAGAATCTGTCAGTTAAGGATATTCGTCGTGGTATGGTTGCTGGTGATGCCAAGAATGATCCCCCTCAAGAGACTGAAGATTTTAATGCCCAAGTTATTATTCTCAATCACCCTGGTCAGATCCATGCAGGATATGCCCCAGTGCTTGATTGTCACACTGCTCATATCGCCTGTAAGTTCAGCGAGATTCTCTCCAAAGTAGATCGTAGATCTGGTCAAGAGACTGAGGCTGCCCCTAAGAATATCAAGAACGGAGATGCCCGCCATAGTTAAACTCACTCCCTCCCAAGGCCCATGTGTGTGGAGTCTTTCTCTGATTACCCACCCCCTTGG'\n",
    "\n",
    "umi_raw = []\n",
    "# to generate 96 forward and 96 reverse umis\n",
    "# and check for nucleotide composition balance\n",
    "# 4**16 is as many combinations as we can get =  4294967296\n",
    "# 100000 is the number of initial random tags, you can reduce or increase the number if it takes too long\n",
    "# or if no enough umis pass the test\n",
    "while len(umi_raw) <= 100000:\n",
    "    umi = create_umi()\n",
    "    umi_pass1 = count_nns(umi, threshold)\n",
    "    \n",
    "    # add the umi to the list if it is not already there\n",
    "    if umi_pass1 not in umi_raw and umi_pass1 != None:\n",
    "        umi_raw.append(umi_pass1)\n",
    "print('Raw UMIs created: %s'%(len(umi_raw)))\n",
    "print(umi_raw[0:10], '\\n')\n",
    "\n",
    "# check umis for homopolymers\n",
    "\n",
    "N = 2 # allow the longest homopolymer to have two nucleotides\n",
    "P = 2 # allow a max of one homopolymer\n",
    "\n",
    "umi_uniq = []\n",
    "\n",
    "for umi in umi_raw:\n",
    "    if check_homopolymers(umi, N, P) is False:\n",
    "        pass\n",
    "    else:\n",
    "        if umi not in umi_uniq:\n",
    "            umi_uniq.append(umi)\n",
    "print('Non-homopolymer UMIs with homopolymer number threshold %s and nucleotide count threshold %s'%(P,N))\n",
    "print('Remaining UMIs %s'%(len(umi_uniq)))\n",
    "print(umi_uniq[0:10], '\\n')\n",
    "\n",
    "# check for matches\n",
    "umi_nofmatch = []\n",
    "\n",
    "for umi in umi_uniq:\n",
    "    check_seq_match(umi, target)\n",
    "    umibool, statement = check_seq_match(umi, target)\n",
    "    if umibool != False: \n",
    "        if umibool not in umi_nofmatch and umibool != True:\n",
    "            umi_nofmatch.append(umibool)\n",
    "print('UMIs with no full match')\n",
    "print('Remaining UMIs %s'%(len(umi_nofmatch)))\n",
    "print(umi_nofmatch[0:10], '\\n')\n",
    "\n",
    "# check for partial matches\n",
    "umi_preselected = []\n",
    "\n",
    "for umi in umi_nofmatch:\n",
    "    umibool, statement = check_partial_matches(umi, target, min_match = 6)\n",
    "    if umibool != False:\n",
    "        if umibool not in umi_preselected:\n",
    "            umi_preselected.append(umibool)\n",
    "            \n",
    "print('UMIs with no partial match')\n",
    "print('Remaining UMIs %s'%(len(umi_preselected))) # number of umis that pass all tests\n",
    "umi_preselected[0:10] # prints the first 10 in the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "afdff837-4aab-41d4-97fa-5a6077ac1cfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "142435\n",
      "Counter({'C': 36644, 'G': 36122, 'A': 35044, 'T': 34625})\n",
      "36644\n"
     ]
    }
   ],
   "source": [
    "# all sequences pooled should start with the same nucleotide to prevent biases during sequencing and amplification\n",
    "# see 10.1038/s41579-018-0116-y and 10.1111/mec.16460\n",
    "# count how many umis you have per nucleotide (first nucleotide)\n",
    "\n",
    "first_nn = [umi[0] for umi in umi_preselected]\n",
    "print(len(first_nn)) # should be the same number of umis in the pre-selected list\n",
    "\n",
    "print(Counter(first_nn))\n",
    "\n",
    "# selecting the nucleotide with the highest counts maximizes the number of tagged primers\n",
    "# that pass the thresholds\n",
    "\n",
    "umis_selected = [umi for umi in umi_preselected if umi[0] == 'C']\n",
    "print(len(umis_selected))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "ef1fbbc2-9f35-41cf-a521-15a419d8a8d5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with forward primers\n",
      "done with reverse primers\n",
      "Umis generated under the thresholds: 192\n",
      "Of which 96 are forward and 96 are reverse\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense</th>\n",
       "      <th>umi_seq</th>\n",
       "      <th>padd_seq</th>\n",
       "      <th>primer_seq</th>\n",
       "      <th>full_primer</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGAAGACTCGCGTCGT</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGAAGACTCGCGTCGTGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.082963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGAGTCTTCAGCGACG</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGAGTCTTCAGCGACGGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.819301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGGCACAGAGTTCTGC</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGGCACAGAGTTCTGCGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.254998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTCGCTGTAGAGCCAG</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CTCGCTGTAGAGCCAGGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.432762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGATCTGGAGTCGACC</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGATCTGGAGTCGACCGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.423528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>forward</td>\n",
       "      <td>CAGAGCGCCAGAGACC</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CAGAGCGCCAGAGACCGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>50.0</td>\n",
       "      <td>66.922446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGCAGCACTGGTTGCA</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGCAGCACTGGTTGCAGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.692364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGGTGTCTACACACGG</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGGTGTCTACACACGGGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.657744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGGTGCACAGTCAGTC</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGGTGCACAGTCAGTCGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>66.078574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGCTCGTGGATCGACA</td>\n",
       "      <td>GGTAG</td>\n",
       "      <td>GAYTTCATCAAGAACATGAT</td>\n",
       "      <td>CGCTCGTGGATCGACAGGTAGGAYTTCATCAAGAACATGAT</td>\n",
       "      <td>47.5</td>\n",
       "      <td>65.855245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sense           umi_seq padd_seq            primer_seq  \\\n",
       "0  forward  CGAAGACTCGCGTCGT    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "1  forward  CGAGTCTTCAGCGACG    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "2  forward  CGGCACAGAGTTCTGC    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "3  forward  CTCGCTGTAGAGCCAG    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "4  forward  CGATCTGGAGTCGACC    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "5  forward  CAGAGCGCCAGAGACC    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "6  forward  CGCAGCACTGGTTGCA    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "7  forward  CGGTGTCTACACACGG    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "8  forward  CGGTGCACAGTCAGTC    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "9  forward  CGCTCGTGGATCGACA    GGTAG  GAYTTCATCAAGAACATGAT   \n",
       "\n",
       "                                 full_primer  gc_content       temp  \n",
       "0  CGAAGACTCGCGTCGTGGTAGGAYTTCATCAAGAACATGAT        47.5  66.082963  \n",
       "1  CGAGTCTTCAGCGACGGGTAGGAYTTCATCAAGAACATGAT        47.5  65.819301  \n",
       "2  CGGCACAGAGTTCTGCGGTAGGAYTTCATCAAGAACATGAT        47.5  66.254998  \n",
       "3  CTCGCTGTAGAGCCAGGGTAGGAYTTCATCAAGAACATGAT        47.5  65.432762  \n",
       "4  CGATCTGGAGTCGACCGGTAGGAYTTCATCAAGAACATGAT        47.5  65.423528  \n",
       "5  CAGAGCGCCAGAGACCGGTAGGAYTTCATCAAGAACATGAT        50.0  66.922446  \n",
       "6  CGCAGCACTGGTTGCAGGTAGGAYTTCATCAAGAACATGAT        47.5  66.692364  \n",
       "7  CGGTGTCTACACACGGGGTAGGAYTTCATCAAGAACATGAT        47.5  65.657744  \n",
       "8  CGGTGCACAGTCAGTCGGTAGGAYTTCATCAAGAACATGAT        47.5  66.078574  \n",
       "9  CGCTCGTGGATCGACAGGTAGGAYTTCATCAAGAACATGAT        47.5  65.855245  "
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finally, calculate GC content and annealing temperature\n",
    "# for the remaining 142435 umis (we need 96*2)\n",
    "# we can check the first 1000 or so umis\n",
    "# we will select those with compatible temperatures and good CG content\n",
    "\n",
    "padd = 'GGTAG'\n",
    "primer_forward = 'GAYTTCATCAAGAACATGAT'\n",
    "primer_reverse = 'GACGTTGAADCCRACRTTGTC'\n",
    "gc_min = 47\n",
    "gc_max = 53\n",
    "# based on https://www.tandfonline.com/doi/full/10.1080/15572536.2006.11832842#d1e421 - FUNGI!\n",
    "# no more than 5 degree difference in annealing temperatures between primer pairs\n",
    "temp_min = 62\n",
    "temp_max = 67\n",
    "\n",
    "# how many tagged primers to generate\n",
    "# e.g. if num_primers = 96, it will generate 96 forward and 96 reverse\n",
    "num_primers = 96\n",
    "\n",
    "# creates the table to store information\n",
    "tagged_primers = pd.DataFrame(columns = ['sense','umi_seq','padd_seq','primer_seq','full_primer','gc_content','temp'])\n",
    "\n",
    "# to make sure no same umi is used for two primers within one pair:\n",
    "# i.e. one umi used in a forward primer cannot longer be used for a forward.\n",
    "umi_track = []\n",
    "\n",
    "# initiate counters to generate just the primers necessary\n",
    "f_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    # padd should be between umi and primer to prevent the umi from acting as a primer\n",
    "    # see 10.1038/s41579-018-0116-y and 10.1111/mec.16460\n",
    "    if umi not in umi_track:\n",
    "        fullseq = umi + padd + primer_forward\n",
    "        # print(fullseq)\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        # print(gc_content)\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        # print(temp)\n",
    "        sense = 'forward' # python is zero indexed\n",
    "\n",
    "        # add thresholds if you wanna keep padd+umi+primer temperatures that overlap the\n",
    "        # PCR protocol annealing temperatures already\n",
    "        # and to keep the GC content close to 50% - second filter for that\n",
    "        # greater than or equal AND lesser then or equal to\n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_forward, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            # add to the counters\n",
    "            f_count += 1\n",
    "            # print('passed %s'%(f_count))\n",
    "            if f_count == num_primers:\n",
    "                break\n",
    "print('done with forward primers')\n",
    "        \n",
    "r_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = umi + padd + primer_reverse\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'reverse' # python is zero indexed\n",
    "        \n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_reverse, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            r_count += 1\n",
    "            if r_count == num_primers:\n",
    "                break\n",
    "    \n",
    "print('done with reverse primers')\n",
    "            \n",
    "print('Umis generated under the thresholds: %s'%(len(tagged_primers)))\n",
    "print('Of which %s are forward and %s are reverse'%(len(tagged_primers[tagged_primers['sense'] == 'forward']),\n",
    "                                                    len(tagged_primers[tagged_primers['sense'] == 'reverse'])))\n",
    "tagged_primers.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "c8d8ae89-7f8f-4114-b860-2e361e876f8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All primers checked\n"
     ]
    }
   ],
   "source": [
    "# before saving the primers, we need to test if they can form hairpin structures that would affect PCR\n",
    "\n",
    "for row in tagged_primers.itertuples():\n",
    "    # print(row.full_primer)\n",
    "    # replace ambiguous nucleotides\n",
    "    for nn in ['A','T','C','G']:\n",
    "        # ambiguous nucleotides include\n",
    "        seq = re.sub('[NRYKMSWBDHV]',nn, row.full_primer, flags = re.IGNORECASE)\n",
    "        # seq = row.full_primer.replace('Y',nn)\n",
    "        # predict secondary structures\n",
    "        structures = fold(seq) # doesn't like ambiguous nucleotides\n",
    "        for struct in structures:\n",
    "            # concerning Delta Gs are the negative ones (indicate estability of the structure)\n",
    "            # also worrying are hairpin structures that are long, two nucleotides is probably fine\n",
    "            # also worrying if the hairpin happens at the 3' end, that affects primer annealing\n",
    "            if 'HAIRPIN' in str(struct.desc) and struct.e < 0:\n",
    "                print(f\"Type: {struct.desc}, Position: {struct.ij}, Delta G: {struct.e:.2f} kcal/mol\")\n",
    "                print(f\"Concerning primer sequence: {row.full_primer.replace}\")\n",
    "print('All primers checked')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "7bf004ce-91ce-4549-af48-ee3777f63755",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file!\n",
    "# TEF1α region Al33F (5′-GAYTTCATCAAGAACATGAT-′3) and Al33R (5′-GACGTTGAADCCRACRTTGTC-′3)\n",
    "tagged_primers.to_csv('./TEFa_Al33F-Al33R_tagged.csv', sep = '\\t', index_label = 'index')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2ecd35e-fa9d-4929-9c8c-2a4527ee6aa3",
   "metadata": {},
   "source": [
    "# ITS metabarcoding for ENTWINE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14f7c52c-2a18-4994-a762-92cfe94268a2",
   "metadata": {},
   "source": [
    "**sampling**\n",
    "\n",
    "9 sites and 6 trees on each site: 54 deadwood logs \n",
    "\n",
    "54 logs + 3 controls (fieldwork, extraction/amplification, sequencing) + sampled 3 times a year for 2 years:  342 deadwood samples to amplify. Give or take, 400 different primer combinations. I need to find two numbers that allow for 400 combinations (to minimise costs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "48e95dad-1b2a-4e2e-999d-853d721e1d7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The numbers 1 and 400 multiplied give 400\n",
      "synthesising a total of 401 primers\n",
      "The numbers 2 and 200 multiplied give 400\n",
      "synthesising a total of 202 primers\n",
      "The numbers 4 and 100 multiplied give 400\n",
      "synthesising a total of 104 primers\n",
      "The numbers 5 and 80 multiplied give 400\n",
      "synthesising a total of 85 primers\n",
      "The numbers 8 and 50 multiplied give 400\n",
      "synthesising a total of 58 primers\n",
      "The numbers 10 and 40 multiplied give 400\n",
      "synthesising a total of 50 primers\n",
      "The numbers 16 and 25 multiplied give 400\n",
      "synthesising a total of 41 primers\n",
      "The numbers 20 and 20 multiplied give 400\n",
      "synthesising a total of 40 primers\n"
     ]
    }
   ],
   "source": [
    "# number of primer combinations that should be unique\n",
    "number = 400\n",
    "\n",
    "# Function to find factor pairs\n",
    "def find_factor_pairs(num):\n",
    "    factor_pairs = []\n",
    "    for i in range(1, int(num**0.5) + 1):\n",
    "        if num % i == 0:\n",
    "            factor_pairs.append((i, num // i))\n",
    "    return factor_pairs\n",
    "\n",
    "factor_pairs = find_factor_pairs(number)\n",
    "\n",
    "for pair in factor_pairs:\n",
    "    print(f\"The numbers {pair[0]} and {pair[1]} multiplied give {number}\")\n",
    "    print(f\"synthesising a total of {pair[0]+pair[1]} primers\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89b07666-cd37-4227-9ff6-ffba944a875b",
   "metadata": {},
   "source": [
    "**primer choice**\n",
    "\n",
    "Based on [Nilsson et al., 2019](10.1038/s41579-018-0116-y) and [Tedersoo et al., 2022](10.1111/mec.16460):\n",
    "\n",
    "I would select ITS9MUNngs because it targets **the whole ITS** region (one and two) and can identify 100% of fungi (although bind to other eukaryotes too). For reverse, I would select LR5F as it doesnt bind to plants and can bind to 99% of fungi. Other primers targeting only fungi can bind to only 95% taxa. If I need an “internal” primer,  I will choose gITS7ngs because targets ITS2 (more recommended than ITS1 by the paper) and can bind to 100% fungi, plus other plants and protists. Reverse can be the same. The problem is that the reverse primer combined with the gITS7ngs has more than 5 5-degree annealing temperature difference which is not recommended. However, choosing ITS4ngUNI (other recommended reverse) has the same 5-degree annealing temperature difference as ITS9MUNngs. I’ll stick to ITS9MUNngs+LR5F for most samples and for anything else failing I will use the gITS7ngs as a backup for ITS2 at least, despite the 5-degree difference.\n",
    "\n",
    "Given that I want to have 400 unique combinations and that I will use the same reverse primer (but two different forward primers), a good option is to go for:\n",
    "\n",
    "- 10 forward and 40 reverse give 400 combinations. With twice as many forward primers (20): synthesising a total of 60 primers  \n",
    "\n",
    "- **16 forward and 25 reverse give 400 combinations. With twice as many forward primers (32): synthesising a total of 57 primers**  \n",
    "\n",
    "- 20 forward and 20 reverse give 400 combinations. With twice as many forward primers (40): synthesising a total of 60 primers  \n",
    "\n",
    "As a backup region, I will choose LSU because it is a longer fragment and is less conserved across fungi than **SSU**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "872f6396-068d-47e7-9b9b-b787eeea2b03",
   "metadata": {},
   "source": [
    "### ITS9MUNngs and LR5F primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "1c39d623-cae9-45a3-bbee-38e4f987f6e5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with forward primers\n",
      "done with reverse primers\n",
      "Umis generated under the thresholds: 50\n",
      "Of which 25 are forward and 25 are reverse\n",
      "All primers checked for hairpin structures\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense</th>\n",
       "      <th>umi_seq</th>\n",
       "      <th>padd_seq</th>\n",
       "      <th>primer_seq</th>\n",
       "      <th>full_primer</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTACACAATCTATACT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTACACAATCTATACTATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.014699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTACACTAATCAATCT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTACACTAATCAATCTATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.311343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTCATATCATATCTAC</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTCATATCATATCTACATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.097630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forward</td>\n",
       "      <td>CATTATACCATCTACA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CATTATACCATCTACAATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.656754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTCACACATCTTATAT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTCACACATCTTATATATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.349482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTATTCATCTACACAT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTATTCATCTACACATATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.349482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>forward</td>\n",
       "      <td>CATCTACTCACATTAA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CATCTACTCACATTAAATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.895517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTATATCCTCTAACAT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTATATCCTCTAACATATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>64.777496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTCCAACTATATCATA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CTCCAACTATATCATAATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.367005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>forward</td>\n",
       "      <td>CACTATTACTCAATAC</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>TACACACCGCCCGTCG</td>\n",
       "      <td>CACTATTACTCAATACATCGCTACACACCGCCCGTCG</td>\n",
       "      <td>51.351351</td>\n",
       "      <td>65.600201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sense           umi_seq padd_seq        primer_seq  \\\n",
       "0  forward  CTACACAATCTATACT    ATCGC  TACACACCGCCCGTCG   \n",
       "1  forward  CTACACTAATCAATCT    ATCGC  TACACACCGCCCGTCG   \n",
       "2  forward  CTCATATCATATCTAC    ATCGC  TACACACCGCCCGTCG   \n",
       "3  forward  CATTATACCATCTACA    ATCGC  TACACACCGCCCGTCG   \n",
       "4  forward  CTCACACATCTTATAT    ATCGC  TACACACCGCCCGTCG   \n",
       "5  forward  CTATTCATCTACACAT    ATCGC  TACACACCGCCCGTCG   \n",
       "6  forward  CATCTACTCACATTAA    ATCGC  TACACACCGCCCGTCG   \n",
       "7  forward  CTATATCCTCTAACAT    ATCGC  TACACACCGCCCGTCG   \n",
       "8  forward  CTCCAACTATATCATA    ATCGC  TACACACCGCCCGTCG   \n",
       "9  forward  CACTATTACTCAATAC    ATCGC  TACACACCGCCCGTCG   \n",
       "\n",
       "                             full_primer  gc_content       temp  \n",
       "0  CTACACAATCTATACTATCGCTACACACCGCCCGTCG   51.351351  65.014699  \n",
       "1  CTACACTAATCAATCTATCGCTACACACCGCCCGTCG   51.351351  65.311343  \n",
       "2  CTCATATCATATCTACATCGCTACACACCGCCCGTCG   51.351351  65.097630  \n",
       "3  CATTATACCATCTACAATCGCTACACACCGCCCGTCG   51.351351  65.656754  \n",
       "4  CTCACACATCTTATATATCGCTACACACCGCCCGTCG   51.351351  65.349482  \n",
       "5  CTATTCATCTACACATATCGCTACACACCGCCCGTCG   51.351351  65.349482  \n",
       "6  CATCTACTCACATTAAATCGCTACACACCGCCCGTCG   51.351351  65.895517  \n",
       "7  CTATATCCTCTAACATATCGCTACACACCGCCCGTCG   51.351351  64.777496  \n",
       "8  CTCCAACTATATCATAATCGCTACACACCGCCCGTCG   51.351351  65.367005  \n",
       "9  CACTATTACTCAATACATCGCTACACACCGCCCGTCG   51.351351  65.600201  "
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finally, calculate GC content and annealing temperature\n",
    "# for the remaining 142435 umis (we need 96*2)\n",
    "# we can check the first 1000 or so umis\n",
    "# we will select those with compatible temperatures and good CG content\n",
    "\n",
    "# blasting this padd against internal transcribed spacer sequences from fungi (core_nt)\n",
    "# return no significant matches (too short of a sequence)\n",
    "\n",
    "padd = 'ATCGC'\n",
    "\n",
    "primer_forward = 'TACACACCGCCCGTCG'\n",
    "primer_reverse = 'CGATCGATTTGCACGTCAGA'\n",
    "\n",
    "gc_min = 46\n",
    "gc_max = 54\n",
    "\n",
    "# no more than 5 degree difference in annealing temperatures between primer pairs\n",
    "temp_min = 62\n",
    "temp_max = 67\n",
    "\n",
    "# how many tagged primers to generate\n",
    "# e.g. if num_primers = 96, it will generate 96 forward and 96 reverse\n",
    "num_primers = 25 # max number, we can select the \"first 16 forward\"\n",
    "\n",
    "# creates the table to store information\n",
    "tagged_primers = pd.DataFrame(columns = ['sense','umi_seq','padd_seq','primer_seq','full_primer','gc_content','temp'])\n",
    "\n",
    "# to make sure no same umi is used for two primers within one pair:\n",
    "# i.e. one umi used in a forward primer cannot longer be used for a forward.\n",
    "umi_track = []\n",
    "\n",
    "# initiate counters to generate just the primers necessary\n",
    "f_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    # padd should be between umi and primer to prevent the umi from acting as a primer\n",
    "    # see 10.1038/s41579-018-0116-y and 10.1111/mec.16460\n",
    "    if umi not in umi_track:\n",
    "        fullseq = umi + padd + primer_forward\n",
    "        # print(fullseq)\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        # print(gc_content)\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        # print(temp)\n",
    "        sense = 'forward' # python is zero indexed\n",
    "\n",
    "        # add thresholds if you wanna keep padd+umi+primer temperatures that overlap the\n",
    "        # PCR protocol annealing temperatures already\n",
    "        # and to keep the GC content close to 50% - second filter for that\n",
    "        # greater than or equal AND lesser then or equal to\n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_forward, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            # add to the counters\n",
    "            f_count += 1\n",
    "            # print('passed %s'%(f_count))\n",
    "            if f_count == num_primers:\n",
    "                break\n",
    "print('done with forward primers')\n",
    "        \n",
    "r_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = padd + umi + primer_reverse\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'reverse' # python is zero indexed\n",
    "        \n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_reverse, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            r_count += 1\n",
    "            if r_count == num_primers:\n",
    "                break\n",
    "    \n",
    "print('done with reverse primers')\n",
    "            \n",
    "print('Umis generated under the thresholds: %s'%(len(tagged_primers)))\n",
    "print('Of which %s are forward and %s are reverse'%(len(tagged_primers[tagged_primers['sense'] == 'forward']),\n",
    "                                                    len(tagged_primers[tagged_primers['sense'] == 'reverse'])))\n",
    "\n",
    "# checking for hairpins\n",
    "for row in tagged_primers.itertuples():\n",
    "    for nn in ['A','T','C','G']:\n",
    "        seq = re.sub('[NRYKMSWBDHV]',nn, row.full_primer, flags = re.IGNORECASE)\n",
    "        structures = fold(seq) # doesn't like ambiguous nucleotides\n",
    "        for struct in structures:\n",
    "            if 'HAIRPIN' in str(struct.desc) and struct.e < 0:\n",
    "                print(f\"Type: {struct.desc}, Position: {struct.ij}, Delta G: {struct.e:.2f} kcal/mol\")\n",
    "                print(f\"Concerning primer sequence: {row.full_primer.replace}\")\n",
    "print('All primers checked for hairpin structures')\n",
    "# save the file\n",
    "tagged_primers.to_csv('./ITS9MUNngs_LR5F-ITSfull_tagged.csv', sep = '\\t', index_label = 'index')\n",
    "\n",
    "tagged_primers.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d773605-cbd2-4296-a2b8-25a29aa6e23b",
   "metadata": {},
   "source": [
    "### gITS7ngs and LR5F primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "1cd8c786-1d1b-4fd7-9404-7553612c8e02",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with forward primers\n",
      "done with reverse primers\n",
      "Umis generated under the thresholds: 50\n",
      "Of which 25 are forward and 25 are reverse\n",
      "All primers checked for hairpin structures\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense</th>\n",
       "      <th>umi_seq</th>\n",
       "      <th>padd_seq</th>\n",
       "      <th>primer_seq</th>\n",
       "      <th>full_primer</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forward</td>\n",
       "      <td>CAGTGTGTCAACACGG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CAGTGTGTCAACACGGATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>64.613421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTCACATGTGGCTGAG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CTCACATGTGGCTGAGATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>64.227087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forward</td>\n",
       "      <td>CATTAGAGGCTCGCTG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CATTAGAGGCTCGCTGATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>64.135071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGGCATCTCTGTGACA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CGGCATCTCTGTGACAATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>64.546286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTGTGACAGCACTCGA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CTGTGACAGCACTCGAATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>64.492471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGTGATCTCCAGGATG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CGTGATCTCCAGGATGATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>63.807157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGAAGACTCGCGTCGT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CGAAGACTCGCGTCGTATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>65.521103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>forward</td>\n",
       "      <td>CCTAGGATCGTAGACG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CCTAGGATCGTAGACGATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>63.178765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>forward</td>\n",
       "      <td>CACGAGTCTACTGGCT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CACGAGTCTACTGGCTATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>63.598425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>forward</td>\n",
       "      <td>CATGAGACACTGCGTC</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GTGARTCATCRARTYTTTG</td>\n",
       "      <td>CATGAGACACTGCGTCATCGCGTGARTCATCRARTYTTTG</td>\n",
       "      <td>47.222222</td>\n",
       "      <td>64.531252</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sense           umi_seq padd_seq           primer_seq  \\\n",
       "0  forward  CAGTGTGTCAACACGG    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "1  forward  CTCACATGTGGCTGAG    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "2  forward  CATTAGAGGCTCGCTG    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "3  forward  CGGCATCTCTGTGACA    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "4  forward  CTGTGACAGCACTCGA    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "5  forward  CGTGATCTCCAGGATG    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "6  forward  CGAAGACTCGCGTCGT    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "7  forward  CCTAGGATCGTAGACG    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "8  forward  CACGAGTCTACTGGCT    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "9  forward  CATGAGACACTGCGTC    ATCGC  GTGARTCATCRARTYTTTG   \n",
       "\n",
       "                                full_primer  gc_content       temp  \n",
       "0  CAGTGTGTCAACACGGATCGCGTGARTCATCRARTYTTTG   47.222222  64.613421  \n",
       "1  CTCACATGTGGCTGAGATCGCGTGARTCATCRARTYTTTG   47.222222  64.227087  \n",
       "2  CATTAGAGGCTCGCTGATCGCGTGARTCATCRARTYTTTG   47.222222  64.135071  \n",
       "3  CGGCATCTCTGTGACAATCGCGTGARTCATCRARTYTTTG   47.222222  64.546286  \n",
       "4  CTGTGACAGCACTCGAATCGCGTGARTCATCRARTYTTTG   47.222222  64.492471  \n",
       "5  CGTGATCTCCAGGATGATCGCGTGARTCATCRARTYTTTG   47.222222  63.807157  \n",
       "6  CGAAGACTCGCGTCGTATCGCGTGARTCATCRARTYTTTG   50.000000  65.521103  \n",
       "7  CCTAGGATCGTAGACGATCGCGTGARTCATCRARTYTTTG   47.222222  63.178765  \n",
       "8  CACGAGTCTACTGGCTATCGCGTGARTCATCRARTYTTTG   47.222222  63.598425  \n",
       "9  CATGAGACACTGCGTCATCGCGTGARTCATCRARTYTTTG   47.222222  64.531252  "
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finally, calculate GC content and annealing temperature\n",
    "# for the remaining 142435 umis (we need 96*2)\n",
    "# we can check the first 1000 or so umis\n",
    "# we will select those with compatible temperatures and good CG content\n",
    "\n",
    "# blasting this padd against internal transcribed spacer sequences from fungi (core_nt)\n",
    "# return no significant matches (too short of a sequence)\n",
    "\n",
    "padd = 'ATCGC'\n",
    "\n",
    "primer_forward = 'GTGARTCATCRARTYTTTG'\n",
    "primer_reverse = 'CGATCGATTTGCACGTCAGA'\n",
    "\n",
    "gc_min = 46\n",
    "gc_max = 54\n",
    "\n",
    "# no more than 5 degree difference in annealing temperatures between primer pairs\n",
    "temp_min = 62\n",
    "temp_max = 67\n",
    "\n",
    "# how many tagged primers to generate\n",
    "# e.g. if num_primers = 96, it will generate 96 forward and 96 reverse\n",
    "num_primers = 25 # max number, we can select the \"first 16 forward\"\n",
    "\n",
    "# creates the table to store information\n",
    "tagged_primers = pd.DataFrame(columns = ['sense','umi_seq','padd_seq','primer_seq','full_primer','gc_content','temp'])\n",
    "\n",
    "# I am not reinitiating this to make sure the tags in the forward primers don't repeat\n",
    "# umi_track = []\n",
    "\n",
    "f_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = umi + padd + primer_forward\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'forward' # python is zero indexed\n",
    "\n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_forward, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            f_count += 1\n",
    "            if f_count == num_primers:\n",
    "                break\n",
    "print('done with forward primers')\n",
    "        \n",
    "r_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = padd + umi + primer_reverse\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'reverse' # python is zero indexed\n",
    "        \n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_reverse, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            r_count += 1\n",
    "            if r_count == num_primers:\n",
    "                break\n",
    "    \n",
    "print('done with reverse primers')\n",
    "            \n",
    "print('Umis generated under the thresholds: %s'%(len(tagged_primers)))\n",
    "print('Of which %s are forward and %s are reverse'%(len(tagged_primers[tagged_primers['sense'] == 'forward']),\n",
    "                                                    len(tagged_primers[tagged_primers['sense'] == 'reverse'])))\n",
    "\n",
    "# checking for hairpins\n",
    "for row in tagged_primers.itertuples():\n",
    "    for nn in ['A','T','C','G']:\n",
    "        seq = re.sub('[NRYKMSWBDHV]',nn, row.full_primer, flags = re.IGNORECASE)\n",
    "        structures = fold(seq) # doesn't like ambiguous nucleotides\n",
    "        for struct in structures:\n",
    "            if 'HAIRPIN' in str(struct.desc) and struct.e < 0:\n",
    "                print(f\"Type: {struct.desc}, Position: {struct.ij}, Delta G: {struct.e:.2f} kcal/mol\")\n",
    "                print(f\"Concerning primer sequence: {row.full_primer.replace}\")\n",
    "print('All primers checked for hairpin structures')\n",
    "\n",
    "# save the file\n",
    "tagged_primers.to_csv('./gITS7ngs_LR5F-ITS2_tagged.csv', sep = '\\t', index_label = 'index')\n",
    "\n",
    "tagged_primers.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cae2f6fb-d03c-4db6-9f20-94ddc81c3b35",
   "metadata": {},
   "source": [
    "### ITS9MUNngs and LR5F primers\n",
    "\n",
    "32\tLSU\tFLR3\tforward\tGGTCCGTGTTTCAAGACG\tLSU\t55\tTedersoo et al., 2015  \n",
    "33\tLSU\tFLR4\treverse\tTGCATGGCCGAATAGGATTA\tLSU\t55\tTedersoo et al., 2015  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "287a21e7-2ee6-4daf-aafa-cbb8c67f2fc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with forward primers\n",
      "done with reverse primers\n",
      "Umis generated under the thresholds: 50\n",
      "Of which 25 are forward and 25 are reverse\n",
      "All primers checked for hairpin structures\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense</th>\n",
       "      <th>umi_seq</th>\n",
       "      <th>padd_seq</th>\n",
       "      <th>primer_seq</th>\n",
       "      <th>full_primer</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTCCTGATAGGATAGA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CTCCTGATAGGATAGAATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.119080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTAGACAATGTGTCTG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CTAGACAATGTGTCTGATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.867787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGCGTAGATACTTAGT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CGCGTAGATACTTAGTATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.498234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTATCTAGTCAAGCAG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CTATCTAGTCAAGCAGATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.494482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forward</td>\n",
       "      <td>CACGAAGAGTCTATTG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CACGAAGAGTCTATTGATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.876192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>forward</td>\n",
       "      <td>CCAATCGTACTGAGTA</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CCAATCGTACTGAGTAATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.888982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTGATACATAGTCGTC</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CTGATACATAGTCGTCATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.634420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>forward</td>\n",
       "      <td>CACTAGAGTCACATTG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CACTAGAGTCACATTGATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.867787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>forward</td>\n",
       "      <td>CGACTATGGGTATACT</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CGACTATGGGTATACTATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.115373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>forward</td>\n",
       "      <td>CTAGTCAAGAGATTCG</td>\n",
       "      <td>ATCGC</td>\n",
       "      <td>GGTCCGTGTTTCAAGACG</td>\n",
       "      <td>CTAGTCAAGAGATTCGATCGCGGTCCGTGTTTCAAGACG</td>\n",
       "      <td>51.282051</td>\n",
       "      <td>66.606563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sense           umi_seq padd_seq          primer_seq  \\\n",
       "0  forward  CTCCTGATAGGATAGA    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "1  forward  CTAGACAATGTGTCTG    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "2  forward  CGCGTAGATACTTAGT    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "3  forward  CTATCTAGTCAAGCAG    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "4  forward  CACGAAGAGTCTATTG    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "5  forward  CCAATCGTACTGAGTA    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "6  forward  CTGATACATAGTCGTC    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "7  forward  CACTAGAGTCACATTG    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "8  forward  CGACTATGGGTATACT    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "9  forward  CTAGTCAAGAGATTCG    ATCGC  GGTCCGTGTTTCAAGACG   \n",
       "\n",
       "                               full_primer  gc_content       temp  \n",
       "0  CTCCTGATAGGATAGAATCGCGGTCCGTGTTTCAAGACG   51.282051  66.119080  \n",
       "1  CTAGACAATGTGTCTGATCGCGGTCCGTGTTTCAAGACG   51.282051  66.867787  \n",
       "2  CGCGTAGATACTTAGTATCGCGGTCCGTGTTTCAAGACG   51.282051  66.498234  \n",
       "3  CTATCTAGTCAAGCAGATCGCGGTCCGTGTTTCAAGACG   51.282051  66.494482  \n",
       "4  CACGAAGAGTCTATTGATCGCGGTCCGTGTTTCAAGACG   51.282051  66.876192  \n",
       "5  CCAATCGTACTGAGTAATCGCGGTCCGTGTTTCAAGACG   51.282051  66.888982  \n",
       "6  CTGATACATAGTCGTCATCGCGGTCCGTGTTTCAAGACG   51.282051  66.634420  \n",
       "7  CACTAGAGTCACATTGATCGCGGTCCGTGTTTCAAGACG   51.282051  66.867787  \n",
       "8  CGACTATGGGTATACTATCGCGGTCCGTGTTTCAAGACG   51.282051  66.115373  \n",
       "9  CTAGTCAAGAGATTCGATCGCGGTCCGTGTTTCAAGACG   51.282051  66.606563  "
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finally, calculate GC content and annealing temperature\n",
    "# for the remaining 142435 umis (we need 96*2)\n",
    "# we can check the first 1000 or so umis\n",
    "# we will select those with compatible temperatures and good CG content\n",
    "\n",
    "# blasting this padd against internal transcribed spacer sequences from fungi (core_nt)\n",
    "# return no significant matches (too short of a sequence)\n",
    "\n",
    "padd = 'ATCGC'\n",
    "\n",
    "primer_forward = 'GGTCCGTGTTTCAAGACG'\n",
    "primer_reverse = 'TGCATGGCCGAATAGGATTA'\n",
    "\n",
    "gc_min = 46\n",
    "gc_max = 54\n",
    "\n",
    "# no more than 5 degree difference in annealing temperatures between primer pairs\n",
    "temp_min = 62\n",
    "temp_max = 67\n",
    "\n",
    "# how many tagged primers to generate\n",
    "# e.g. if num_primers = 96, it will generate 96 forward and 96 reverse\n",
    "num_primers = 25 # we are not mixing these up, we can do 16 and 25\n",
    "\n",
    "# creates the table to store information\n",
    "tagged_primers = pd.DataFrame(columns = ['sense','umi_seq','padd_seq','primer_seq','full_primer','gc_content','temp'])\n",
    "\n",
    "f_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = umi + padd + primer_forward\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'forward' # python is zero indexed\n",
    "\n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_forward, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            f_count += 1\n",
    "            if f_count == num_primers:\n",
    "                break\n",
    "print('done with forward primers')\n",
    "        \n",
    "r_count = 0\n",
    "for umi in umis_selected: # umi_preselected[0:20000] if you want to test things fast, you can select the first e.g. 20000 rows\n",
    "    if umi not in umi_track:\n",
    "        fullseq = padd + umi + primer_reverse\n",
    "        gc_content = gc_fraction(fullseq) * 100\n",
    "        temp = mt.Tm_NN(fullseq)\n",
    "        sense = 'reverse' # python is zero indexed\n",
    "        \n",
    "        if gc_min <= gc_content <= gc_max and temp_min <= temp <= temp_max:\n",
    "            row = [sense, umi, padd, primer_reverse, fullseq, gc_content, temp]\n",
    "            tagged_primers.loc[len(tagged_primers)] = row # appends the row\n",
    "            umi_track.append(umi)\n",
    "            r_count += 1\n",
    "            if r_count == num_primers:\n",
    "                break\n",
    "    \n",
    "print('done with reverse primers')\n",
    "            \n",
    "print('Umis generated under the thresholds: %s'%(len(tagged_primers)))\n",
    "print('Of which %s are forward and %s are reverse'%(len(tagged_primers[tagged_primers['sense'] == 'forward']),\n",
    "                                                    len(tagged_primers[tagged_primers['sense'] == 'reverse'])))\n",
    "\n",
    "# checking for hairpins\n",
    "for row in tagged_primers.itertuples():\n",
    "    for nn in ['A','T','C','G']:\n",
    "        seq = re.sub('[NRYKMSWBDHV]',nn, row.full_primer, flags = re.IGNORECASE)\n",
    "        structures = fold(seq) # doesn't like ambiguous nucleotides\n",
    "        for struct in structures:\n",
    "            if 'HAIRPIN' in str(struct.desc) and struct.e < 0:\n",
    "                print(f\"Type: {struct.desc}, Position: {struct.ij}, Delta G: {struct.e:.2f} kcal/mol\")\n",
    "                print(f\"Concerning primer sequence: {row.full_primer.replace}\")\n",
    "print('All primers checked for hairpin structures')\n",
    "\n",
    "# save the file\n",
    "tagged_primers.to_csv('./FLR3_FLR4-LSU_tagged.csv', sep = '\\t', index_label = 'index')\n",
    "\n",
    "tagged_primers.head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
